src/include/platform/NetworkCommissioning.h

```diff
@@ -0,0 +1,189 @@
+#include <lib/core/CHIPSafeCasts.h>
+#include <lib/support/ThreadOperationalDataset.h>
+#include <platform/internal/DeviceNetworkInfo.h>
+
+namespace chip {
+namespace DeviceLayer {
+
+class DLL_EXPORT WiFiNetworkCommissioningDelegate
+{
+public:
+    struct WiFiNetworkInfo
+    {
+        uint8_t mSSID[DeviceLayer::Internal::kMaxWiFiSSIDLength + 1];
+        uint8_t mSSIDLen;
+        uint8_t mCredentials[DeviceLayer::Internal::kMaxWiFiKeyLength];
+        uint8_t mCredentialsLen;
+    };
+
+    class Callback
+    {
+    public:
+        virtual void OnConnectWiFiResult(app::Clusters::NetworkCommissioning::NetworkCommissioningStatus commissioningError,
+                                         CharSpan debugText, int32_t wifiAssociationStatus) = 0;
+
+        /**
+         * Indicates the scan is finished, and accepts a list of networks discovered.
+         * OnScanFinished() must be called in a thread-safe manner with CHIP stack. (e.g. using ScheduleWork or ScheduleLambda)
+         */
+        virtual void
+        OnScanFinished(CHIP_ERROR err, CharSpan debugText,
+                       const Span<app::Clusters::NetworkCommissioning::Structs::WiFiInterfaceScanResult::Type> & network) = 0;
+
+        virtual ~Callback() = default;
+    };
+
+    /**
+     * @brief
+     *   Called to initialtiate a WiFi network scan, exactly one of callback->OnError or callback->OnScanFinished must be called
+     * within timeout.
+     *   The ScanWiFiNetworks has no return values, platforms can call callback->OnError() inside ScanWiFiNetworks.
+     *
+     * @param timeout The timeout of the scan, the scan MUST be finished or terminated within the given timeout.
+     * @param ssid    The interested SSID, the scanning MAY be restricted to to the given SSID.
+     */
+    virtual void ConnectWiFiNetwork(ByteSpan ssid, ByteSpan credentials, Callback * callback) = 0;
+
+    /**
+     * @brief
+     *   Called to initialtiate a WiFi network scan, exactly one of callback->OnError or callback->OnScanFinished must be called
+     * within timeout.
+     *   The ScanWiFiNetworks has no return values, platforms can call callback->OnError() inside ScanWiFiNetworks.
+     *
+     * @param timeout The timeout of the scan, the scan MUST be finished or terminated within the given timeout.
+     * @param ssid    The interested SSID, the scanning MAY be restricted to to the given SSID.
+     */
+    virtual void ScanWiFiNetworks(ByteSpan ssid, Callback * callback) = 0;
+
+    /**
+     * @brief
+     *   Called to retrieve a list of WiFi networks configured. The implementation should fill no more than maxNetworks items in
+     * the networks.
+     *
+     * @param networks The buffer for filling WiFi network credentials.
+     * @param maxNetworks The size of the WiFi network info buffer.
+     * @return The actual number of networks retrieved from platform driver.
+     */
+    virtual size_t GetNetworks(WiFiNetworkInfo * networks, size_t maxNetworks) = 0;
```

Comment:
Suggest moving to an iterator. The iterator can easily be used to do per-index access of the `Networks` list in the implementation of the cluster, since the number will be small. The iterator starts at index zero. The iterator can have a Count() method.

Again, see `src/credentials/GroupDataProvider.h` for a good example.

```
    /**
     * Template used to iterate the stored group data
     */
    template <typename T>
    class Iterator
    {
    public:
        virtual ~Iterator() = default;
        /**
         *  @retval The number of entries in total that will be iterated.
         */
        virtual size_t Count() = 0;
        /**
         *   @param[out] item  Value associated with the next element in the iteration.
         *  @retval true if the next entry is successfully retrieved.
         *  @retval false if no more entries can be found.
         */
        virtual bool Next(T & item) = 0;
        /**
         * Release the memory allocated by this iterator.
         * Must be called before the pointer goes out of scope.
         */
        virtual void Release() = 0;

    protected:
        Iterator() = default;
    };

    using GroupInfoIterator = Iterator<GroupInfo>;
```